src/arch/ssse3.rs

```diff
@@ -0,0 +1,289 @@
+        _mm_subs_epi16(_mm_subs_epi16(p5, p2), p2),
+        _mm_mulhrs_epi16(p2, _mm_set1_epi16(18446)), // 0.562915447
+    );
+
+    let p3 = _mm_subs_epi16(
+        _mm_mulhrs_epi16(p3, _mm_set1_epi16(-31509)), // -0.961570560
+        p3,
+    );
+    let p4 = _mm_mulhrs_epi16(p4, _mm_set1_epi16(-12785)); // -0.390180644
+
+    let t3 = _mm_adds_epi16(_mm_adds_epi16(p1, p4), t3);
+    let t2 = _mm_adds_epi16(_mm_adds_epi16(p2, p3), t2);
+    let t1 = _mm_adds_epi16(_mm_adds_epi16(p2, p4), t1);
+    let t0 = _mm_adds_epi16(_mm_adds_epi16(p1, p3), t0);
+
+    data[0] = _mm_adds_epi16(x0, t3);
+    data[7] = _mm_subs_epi16(x0, t3);
+    data[1] = _mm_adds_epi16(x1, t2);
+    data[6] = _mm_subs_epi16(x1, t2);
+    data[2] = _mm_adds_epi16(x2, t1);
+    data[5] = _mm_subs_epi16(x2, t1);
+    data[3] = _mm_adds_epi16(x3, t0);
+    data[4] = _mm_subs_epi16(x3, t0);
+}
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+#[target_feature(enable = "ssse3")]
+unsafe fn transpose8(data: &mut [__m128i; 8]) {
+    // Transpose a 8x8 matrix with a sequence of interleaving operations.
+    // Naming: dABl contains elements from the *l*ower halves of vectors A and B, interleaved, i.e.
+    // A0 B0 A1 B1 ...
+    // dABCDll contains elements from the lower quarter (ll) of vectors A, B, C, D, interleaved -
+    // A0 B0 C0 D0 A1 B1 C1 D1 ...
+    let d01l = _mm_unpacklo_epi16(data[0], data[1]);
+    let d23l = _mm_unpacklo_epi16(data[2], data[3]);
+    let d45l = _mm_unpacklo_epi16(data[4], data[5]);
+    let d67l = _mm_unpacklo_epi16(data[6], data[7]);
+    let d01h = _mm_unpackhi_epi16(data[0], data[1]);
+    let d23h = _mm_unpackhi_epi16(data[2], data[3]);
+    let d45h = _mm_unpackhi_epi16(data[4], data[5]);
+    let d67h = _mm_unpackhi_epi16(data[6], data[7]);
+    // Operating on 32-bits will interleave *consecutive pairs* of 16-bit integers.
+    let d0123ll = _mm_unpacklo_epi32(d01l, d23l);
+    let d0123lh = _mm_unpackhi_epi32(d01l, d23l);
+    let d4567ll = _mm_unpacklo_epi32(d45l, d67l);
+    let d4567lh = _mm_unpackhi_epi32(d45l, d67l);
+    let d0123hl = _mm_unpacklo_epi32(d01h, d23h);
+    let d0123hh = _mm_unpackhi_epi32(d01h, d23h);
+    let d4567hl = _mm_unpacklo_epi32(d45h, d67h);
+    let d4567hh = _mm_unpackhi_epi32(d45h, d67h);
+    // Operating on 64-bits will interleave *consecutive quadruples* of 16-bit integers.
+    data[0] = _mm_unpacklo_epi64(d0123ll, d4567ll);
+    data[1] = _mm_unpackhi_epi64(d0123ll, d4567ll);
+    data[2] = _mm_unpacklo_epi64(d0123lh, d4567lh);
+    data[3] = _mm_unpackhi_epi64(d0123lh, d4567lh);
+    data[4] = _mm_unpacklo_epi64(d0123hl, d4567hl);
+    data[5] = _mm_unpackhi_epi64(d0123hl, d4567hl);
+    data[6] = _mm_unpacklo_epi64(d0123hh, d4567hh);
+    data[7] = _mm_unpackhi_epi64(d0123hh, d4567hh);
+}
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+#[target_feature(enable = "ssse3")]
+pub unsafe fn dequantize_and_idct_block_8x8(
+    coefficients: &[i16],
```

Comment:
Drive-by review: just like #215 replaced some slices with arrays it would make sense to do the same here